File: JavaStackImplementationBMW/src/StackImp.java
import java.util.Stack;

/**
 * Class for implementing stack data structure of any type
 *
 * @author Raphasha Donald Mailula
 * @version 1.0
 */
public class StackImp{

    Stack<Object> stackElements = new Stack<>();

    /**
     * Creates an empty Stack.
     */
    public StackImp() {
    }

    /**
     * Adding element on the top of the stack
     *
     * @param o the integer to add
     */
    public void push(Object o){
        stackElements.add(count(), o);
    }

    /**
     * Retrieving element from the stack on the iterator index
     *
     * @return the added object
     */
    public Object pop(){
        if(!isEmpty()) {
            int size = count();
            Object e = null;
            try {
                e = stackElements.get(--size);
                stackElements.remove(e);
                return e;
            } catch (ArrayIndexOutOfBoundsException ex){
                return ex.getClass();
            }
        }
        return null;
    }

    /**
     * Checking if the stack has any values
     *
     * @return true if the stack is empty using the count implementation of
     * false inverse
     */
    public boolean isEmpty(){
        return count() == 0;
    }

    /**
     * Counting the number of values on the stack
     *
     * @return integer value of the stack size
     */
    public int count(){
        return stackElements.size();
    }
}

File: TypescriptStackImplementationBMW/src/app/stack/stackImp.ts
/**
  * Class for data structure implementation of the 
  * methods on IStack interface
  * @author Raphasha Donald Mailula
  * @version 1.0
  */
import { IStackImp } from './IStackImp';

export class StackImp<T> implements IStackImp<T> {
    private elements: T[] = [];
  
    /**
      * Creates empty constructor of the Object and setting the 
      * stack capacity to infite
      */
    constructor(private capacity: number = Infinity) {}

    /**
      * Appends new elements to an array, and returns the new length of the array.
      * @param item New element of the Array.
      */
    push(item: T): void {
      if (this.count() === this.capacity) {
        throw Error("Cannot add more items, capacity reached");
      }
      this.elements.push(item);
    }

    /**
      * Checks if the array has values or returns null
      * Removes the last element from an array.
      * @returns the last element from an array
      */
    pop(): T | any {
        if(!this.empty()){
            return this.elements.pop();
        }
        
        return null;
    }

    /**
      * Checks if the array has elements inside
      * @returns True or False
      */
    empty(): T | boolean {
      return this.count() == 0;
    }
  
    /**
      * Checks the size of the array
      * @returns True or False
      */
    count(): number {
      return this.elements.length;
    }
}

  